### PR TITLE
clustermesh: fix a few issues in the new mcs api service controller

### PR DESCRIPTION
**Pull request by MrFreezeex**
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

This commit does 3 small fixes:
- Use the correct upstream MCS-API controller. The controller used
  is now the ones that sync the service IP to the ServiceImport resources.
  The rest of the controllers are Cilium specific and will (or already is)
  be implemented soon.
- Also add a shortcut on creation to save a delete/recreate on
  of the derived service if there is no ServiceImport and the local is
  headless.
- Fix the watch on Services to also issue a reconcile if the locally
  exported Service has changed

Related to #27902 

```release-note
Fix a few issues with the newly added MCS-API controllers
```

### CODE REVIEW

**Comment by YutaroHayakawa** [comment]
CC: @giorio94

File: pkg/annotation/k8s.go [hunk 1]
```diff
@@ -105,6 +105,10 @@ const (
 	SharedService      = ServicePrefix + "/shared"
 	SharedServiceAlias = Prefix + "/shared-service"
 
+	// MCSAPIServiceSpec is an internal annotation to share MCS-API spec to
+	// facilitate the synchronization of MCS ServiceImport across clusters.
+	MCSAPIServiceSpec = ServicePrefix + "/mcs-api-internal-svc-spec"
```
Comment (YutaroHayakawa) [inline pkg/annotation/k8s.go]: I would introduce a new prefix for the internal annotations like "service.internal.cilium.io` or `service-internal.cilium.io` so that users can easily recognize it is not something they should be aware of. WDYT @giorio94?

**Review by YutaroHayakawa — CHANGES_REQUESTED** [review]
Could you elaborate more about how this annotation is used within the service importing/exporting flow? Currently it is very hard to follow it.

File: pkg/clustermesh/mcsapi/service_controller_test.go
Comment (YutaroHayakawa) [inline pkg/clustermesh/mcsapi/service_controller_test.go:1]: I noticed that the name of the test is `Test_httpRouteReconciler_Reconcile`. I guess you copy the code from GWAPI code and forgot to rename?

Comment (MrFreezeex) [reply pkg/clustermesh/mcsapi/service_controller_test.go:1]: Ahh yes, I admit that's my go to in term of reference for controller tests x). Will fix that 👍

File: pkg/service/store/store.go [hunk 2]
```diff
@@ -165,6 +171,135 @@ func NewClusterService(name, namespace string) ClusterService {
...
+	SessionAffinityConfig *v1.SessionAffinityConfig `json:"sessionAffinityConfig,omitempty"`
+}
+
+// Marshal returns the MCS-API Service Spec object as JSON byte slice
+func (s *MCSAPIServiceSpec) Marshal() ([]byte, error) {
+	return json.Marshal(s)
+}
+
+// Unmarshal parses the JSON byte slice and updates the MCS-API Service Spec receiver
+func (s *MCSAPIServiceSpec) Unmarshal(_ string, data []byte) error {
```
Comment (joamaki) [inline pkg/service/store/store.go]: What's this `_ string` about?

File: pkg/clustermesh/mcsapi/service_controller.go [hunk 3]
```diff
@@ -236,6 +236,12 @@ func (r *mcsAPIServiceReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	if localSvc != nil {
 		svc.Spec.Selector = localSvc.Spec.Selector
 		svc.Spec.Ports = localSvc.Spec.Ports
+
+		// Use the local Service on creation to save up a potential round trip of
```
Comment (joamaki) [inline pkg/clustermesh/mcsapi/service_controller.go]: I don't quite understand the relationship with the comment and the code. How does setting the ClusterIP save a "roundtrip"? What was the original reason for adding this?

File: pkg/service/store/store.go [hunk 4]
```diff
@@ -165,6 +171,135 @@ func NewClusterService(name, namespace string) ClusterService {
...
+	// created. It is not guaranteed to be set in happens-before order across separate operations.
+	// Clients may not set this value. It is represented in RFC3339 form and is in UTC.
+	//
+	// Populated by the system.
+	// Read-only.
+	// Null for lists.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	ExportCreationTimestamp metav1.Time `json:"exportCreationTimestamp"`
+	// +listType=atomic
+	Ports []mcsapiv1alpha1.ServicePort `json:"ports"`
```
Comment (joamaki) [inline pkg/service/store/store.go]: Embedding k8s objects into our kvstore json-marshalled backwards compatible objects seems a bit scary to me. We also do want to avoid too many k8s specific things at this layer to somewhat keep the path open to alternative systems in the future. It's also hard to judge how well space-optimized this is. @giorio94 I guess you're our kvstore guy. What do you think about how the `ClusterService` format should evolve and what our rules around it should be? My instinct is that we should keep these structs very minimal and very self-contained.

A related note is that I find it fairly scary that our kvstore objects aren't versioned at all. I wonder if we should at some point think about having a proper schema for all kvstore objects that we can validate? Perhaps going via protobuf or CUE...

File: pkg/service/store/store.go [hunk 5]
```diff
@@ -165,6 +171,135 @@ func NewClusterService(name, namespace string) ClusterService {
...
+	*s = newMCSAPIServiceSpec
+
+	return nil
+}
+
+// DeepEqual is a deepequal function, deeply comparing the receiver with other.
+// in must be non-nil. This is not auto-generated because the auto-generated
+// function doesn't compile because of some fields that we don't control.
+// Although the auto-generated version was taken as a base.
+func (in *MCSAPIServiceSpec) DeepEqual(other *MCSAPIServiceSpec) bool {
```
Comment (joamaki) [inline pkg/service/store/store.go]: I guess we want to remove the `deepcopy-gen=true` above?
Related to my comment about embedding outside k8s objects. Wouldn't it be better if `MCSAPIServiceSpec` would be standalone and we'd be able to generate the `DeepEqual` for it?

File: pkg/clustermesh/mcsapi/service_controller.go [hunk 6]
```diff
@@ -287,21 +293,25 @@ func (r *mcsAPIServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 			if !ok {
 				return []ctrl.Request{}
 			}
-			name := types.NamespacedName{Name: svcExport.Name, Namespace: svcExport.Namespace}
-			return []ctrl.Request{{NamespacedName: name}}
+			return []ctrl.Request{{NamespacedName: types.NamespacedName{
+				Name: svcExport.Name, Namespace: svcExport.Namespace,
+			}}}
```
Comment (giorio94) [inline pkg/clustermesh/mcsapi/service_controller.go]: Drive-by comment. This watches block could be simplified with:

```
Watches(&mcsapiv1alpha1.ServiceExport{}, &handler.EnqueueRequestForObject{}).
```

File: pkg/clustermesh/mcsapi/service_controller.go [hunk 7]
```diff
@@ -287,21 +293,25 @@ func (r *mcsAPIServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 		})).
-		// Watch for changes to Services supported that have a MCS API object as owner reference
+		// Watch for changes to Services
 		Watches(&corev1.Service{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []ctrl.Request {
 			svc, ok := obj.(*corev1.Service)
 			if !ok {
 				return []ctrl.Request{}
 			}
```
Comment (giorio94) [inline pkg/clustermesh/mcsapi/service_controller.go]: Drive-by comment. The cast is not necessary, as all information is available through the `Object` interface.

**Comment by YutaroHayakawa** [comment]
@MrFreezeex Hi, thanks for your clarification!

> As the derived service is already the aggregated/conflict resolved version we need to somehow have a way to find which cluster has what property before aggregating them, this is maybe a bit unclear in this PR because this part is not present yet. The aggregation/conflict resolution will be in another controller whose job is to create ServiceImport.

Ok, then my next question is we could get this disaggregated information from ServiceExport + the actual exported service. Then why do we need to introduce this new annotation?

**Comment by giorio94** [comment]
> Ok, then my next question is we could get this disaggregated information from ServiceExport + the actual exported service. Then why do we need to introduce this new annotation?

I second Yutaro's comment, here. It would be nice to have a diagram (maybe linked to the original issue) which depicts the flows you are thinking about, both in terms of what users would do, and what Cilium will automatically do. That would help a lot in being able to properly review this PR (and the follow-ups), as we are otherwise missing the context about the next steps, and it becomes difficult to provide useful comments.

File: pkg/service/store/store.go (hunk 4 above)
Comment (giorio94) [reply pkg/service/store/store.go]: I agree here with Jussi's comment that the kvstore structs should be maintained as simple as possible, and embedding structs that we don't control feels pretty scary. 

It is also not totally clear to me whether we really need all the fields that are introduced here (e.g., the timestamp), and what would be their use-case. Would the idea be to automatically create `ServiceImport` objects in the target clusters out of this data? In that case it may be worth exploring the idea of storing only the information in a separate kvstore path, as we would be deviating from the original purpose of the ClusterService structure (for instance information about port mappings is already present, but laid out in a different format). That way, we would also reduce the risks of introducing backward-incompatible changes affecting existing functionalities. In any case, a CFP-like discussion would really help IMO.

> A related note is that I find it fairly scary that our kvstore objects aren't versioned at all.

Technically speaking, kvstore paths are versioned (and all at `v1`), but we never explored the possibility of supporting multiple versions in parallel till now AFAIK.

File: pkg/annotation/k8s.go (hunk 1 above)
Comment (giorio94) [reply pkg/annotation/k8s.go]: The suggestion sounds reasonable to me. That said, as you also mentioned below, it is not clear to me at the moment why we need this annotation, rather than having the appropriate controller looking up the corresponding `ServiceImport` object. Storing JSON content in an annotation feels pretty fragile to me.

File: pkg/service/store/store.go (hunk 2 above)
Comment (MrFreezeex) [reply pkg/service/store/store.go]: Ah I took it from `ClusterService` which was doing the same thing but it doesn't seems needed indeed

File: pkg/clustermesh/mcsapi/service_controller.go (hunk 3 above)
Comment (MrFreezeex) [reply pkg/clustermesh/mcsapi/service_controller.go]: So this is triggered when the `ServiceImport` object is not yet created and you only have a `ServiceExport`.

Previously in that case the Service was always created as non headless which means that if the `ServiceImport` that would most likely be created right after (by a controller that doesn't exist yet) would retrigger a reconcile of this controller and if it is headless the "derived service" would then be deleted and recreated.

With this new condition we rely on the local service to determine the headlessness of the derived service so that we might save up the deletion+recreation/not have to switch the headlessness. This however won't be the case if there is a conflict on this property though.

I will update the comment to be a bit more explicit there.

File: pkg/service/store/store.go (hunk 5 above)
Comment (MrFreezeex) [reply pkg/service/store/store.go]: > I guess we want to remove the deepcopy-gen=true above?

The deepcopy generation is necessary for the deepcopy of the `ClusterService` struct to work. And for the DeepEqual there is 3 issues with the generated code:
- It tries to call `DeepEqual` on the `ExportCreationTimestamp` field which doesn't exist (which for reference is a `metav1.Time` object)
- It tries to call `DeepEqual` on each `mcsapiv1alpha1.ServicePort` object which doesn't exists
- It tries to call `DeepEqual` on the `SessionAffinityConfig` (which for reference is a `corev1.SessionAffinityConfig` object)

I am not sure how to resolve those without having to implement the function without the code generation unfortunately :/

**Comment by MrFreezeex** [comment]
@giorio94 @YutaroHayakawa @joamaki thanks for the feedback, I will try to create a dedicated CFP issue with some schema of what the `ServiceImport` auto creation is doing in more details and present a few other alternatives so that we can choose one that make sense with all the info at hands :+1:.

**Comment by MrFreezeex** [comment]
Hi! Since we are having a separate discussion for how data should be shared across clusters to allow automatic ServiceImport creation I dropped the second commit about that and only kept the first one which brought a few fixes. I wanted to keep same PR so that the assigned reviewers don't change (although now it should probably be a sig-clustermesh only change), hopefully that's ok :pray:.

**Review by YutaroHayakawa — APPROVED** [review]

**Review by giorio94 — APPROVED** [review]
These cleanup changes look good to me, thanks!

And sorry for the delay in reviewing the CFP, I plan to share my comments by today.

**Comment by giorio94** [comment]
Jussi's concern has been addressed, and his review is no longer strictly required, given that all changes are clustermesh-related only now. Marking as ready to merge to unblock.
